로그아웃 기능 개발

Habist opened the pull request:
## 작업 내용
- 로그인 상태인지 확인하는 인터셉터 추가 및 등록
- Redis에 저장된 세션 데이터 삭제로 로그아웃 기능 구현

## 주의 사항

>  참조이슈
> #13

Habist committed:
feat: [#13] 로그아웃 기능 개발

로그아웃 시 Redis에 저장하고 있는 세션 정보 삭제하도록 구현했습니다.

Habist committed:
improve: [#13] 로그아웃 기능 수정

로그아웃 전에 로그인 상태인지 확인하는 부분을 제거했습니다.

Habist committed:
feat: [#13] 로그인 상태 확인 인터셉터 추가

로그인 상태인지 확인하는 인터셉터를 추가했습니다.
회원가입과 로그인을 위한 요청을 제외한 모든 요청은 로그인이 필요하기 때문에
exclude 경로를 설정하였습니다.

Habist committed:
refactor: [#13] 로그인 상태 확인 인터셉터 에러 반환 코드 상수로 변경

로그인 상태가 아닐 시 상태코드 상수를 반환하도록 변경하였습니다.

f-lab-dev commented on src/main/java/com/campool/configuration/WebConfiguration.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public void addInterceptors(InterceptorRegistry registry) {
+        registry.addInterceptor(loginCheckInterceptor)
```
필터나 AOP가 아닌 인터셉터로 구현하신 이유가 있으신가요?

Habist replied:
먼저 Filter를 사용하지 않은 이유는
필터는 스프링 영역을 벗어나 모든 서블릿 요청을 가장 앞단에서 받아 처리합니다. 스프링 빈을 사용할 수 없고 다른 필터나 스프링에 의해 걸러질 수 있는 요청들도 전부 받기 때문에 성능적 부분을 고려해 배제하였습니다.

AOP가 아닌 인터셉터로 구현한 이유는
인터셉터를 사용하면 좀더 범용적으로, 쉽게 사용할 수 있다고 생각했습니다. Campool은 회원가입과 로그인을 제외한 모든 요청에는 로그인 인증이 필요한 서비스입니다. AOP로 구현하면 로그인 AOP를 만들고 각 컨트롤러 메소드마다 애노테이션을 추가하는 방식으로 적용해야 하지만 인터셉터를 이용하면 해당 소스와 같이 제외된 요청 경로만 추가해주면 나머지 모든 요청에 적용할 수 있어 훨씬 간편하게 구현할 수 있다고 생각했습니다.

결론적으로 인터셉터로 구현하면 URL에 대한 처리는 가능하지만 #14 처럼 REST API 디자인 원칙을 지키면 동일한 URL에 다른 메소드에 대해서는 제어가 어려운 단점이 발생하고 있는 거 같습니다. 또 인터셉터는 http 요청과 응답에 대해서만 처리가 가능해 세부적인 제어가 필요하거나 컨트롤러 요청이 아닌 메소드 요청에 로그인이 필요한 경우에는 AOP가 좀 더 합리적인 선택이라 느껴지네요.

제가 좀 더 고려해야 할 부분이 있을까요? 없으면 AOP로 다시 구현해볼까 합니다.

f-lab-dev replied:
설명은 합리적인것 같네요~ 비즈니스 요구사항이랑 접목도 잘 해주셨습니다~

Habist committed:
improve: [#13] 로그인 확인 기능 AOP로 변경

인터셉터를 통해 수행되던 로그인 확인 기능을 AOP로 변경하였습니다.
인터셉터는 같은 요청 URL인 경우 METHOD 별 세부적 제어가 어렵고
AOP는 특정 메소드에 개별적인 적용이 훨씬 용이합니다.

f-lab-dev commented on src/main/java/com/campool/aop/LoginValidationAspect.java:
```diff
@@ -0,0 +1,27 @@
+    @Before("@annotation(com.campool.annotation.LoginValidation)")
+    public void loginValidate(JoinPoint joinPoint) {
+        if (!authService.isValidAuthentication()) {
+            throw new HttpClientErrorException(HttpStatus.FORBIDDEN, "로그인이 필요한 서비스입니다.");
```
권한이 없는 것인지, 로그인이 안됀 것인지 구분이 되면 좋을 것 같습니다~

Habist committed:
회원정보 수정 기능 개발 (#17)

* feat: [#14] 회원정보 수정 기능 개발

회원정보 수정 요청에 해당하는 도메인 클래스를 추가했습니다.
인증 서비스에 인증된 사용자 아이디 사용을 위한 메소드를 추가했습니다.
사용자 서비스에 사용자 아이디와 변경 요청을 기반으로 사용자 정보를 수정하는 메소드를 추가했습니다.
사용자 아이디에 해당하는 정보를 변경하는 쿼리를 추가했습니다.

* improve: [#14 - 리뷰 반영] REST API 디자인 준수 및 인증 서비스 메소드명 수정

회원정보 수정하는 요청 URL을 REST API 디자인에 더 맞게 준수하는 쪽으로 변경하였습니다.
인증 서비스 메소드 중 인증된 사용자 아이디를 가져오는 메소드명을 좀 더 명확한 의미로 변경하였습니다.

* feat: [#14 - 리뷰 반영] 현재 로그인 유저 아이디 주입 애노테이션, AOP 추가

현재 세션에 저장되어 있는 사용자 아이디를 주입받을 수 있도록 명시하는 애노테이션을 추가하였습니다.
현재 세션에 저장되어 있는 사용자 아이디를 주입해주는 AOP를 추가하였습니다.

* improve: [#14 - 리뷰 반영] 현재 사용자 아이디 주입하는 AOP Spring Resolver 기반으로 변경

현재 세션에 저장되어 있는 사용자 아이디를 주입해주는 AOP를 제거하고
Spring HandlerMethodArgumentResolver 기반으로 변경하였습니다.

* improve: [#14 - 리뷰 반영] PresentUserIdResolver String 타입 확인 로직 추가

타입 확인을 위해 String 이외의 타입이 들어오면 예외를 발생하도록 변경하였습니다.

* fix: [#14] PresentUserId 파라미터 타입 체크 로직 버그 수정

애노테이션이 없는 매개 변수에도 타입 체크를 하는 버그를 수정하였습니다.
-> PresentUserId 애노테이션이 있는 파라미터만 String 타입 체크를 하도록 변경
